cpp/src/arrow/util/future.h

```diff
@@ -56,6 +47,47 @@ struct is_future<Future<T>> : std::true_type {};
 template <typename Signature>
 using result_of_t = typename std::result_of<Signature>::type;
 
+template <typename Source, typename Dest>
+typename std::enable_if<Source::is_empty>::type Propagate(Source* source, Dest dest) {
+  struct MarkNextFinished {
+    void operator()(const Status& status) && { next.MarkFinished(status); }
+    Dest next;
+  };
+  source->AddCallback(MarkNextFinished{std::move(dest)});
+}
+
+template <typename Source, typename Dest, bool SourceEmpty = Source::is_empty,
+          bool DestEmpty = Dest::is_empty>
+struct MarkNextFinished {};
+
+template <typename Source, typename Dest>
+struct MarkNextFinished<Source, Dest, true, false> {
+  void operator()(const Status& status) && { next.MarkFinished(status); }
```

Comment:
Is this one valid? If the status is ok, then `next` needs an actual `Result` object.

Reply:
Correct.  I've removed this method.